# Psychoinformatics - Week 11 (Exercises)
by 楊宗樺 (b05705050@ntu.edu.tw)

In [2]:
%config IPCompleter.greedy=True 
%matplotlib inline

from numpy import *
from matplotlib.pyplot import *
from IPython.display import *
import warnings
warnings.simplefilter('ignore', DeprecationWarning)
from sklearn import *

iris = datasets.load_iris()
X=iris.data; Y=iris.target

## 1 Performance Tuning of a Neural Net (6 points)

### 1.0 Baseline Performance
SVM can reach an classifcation accuracy ~ 8x% correct for the hard Iris problem.

In [66]:
sss=model_selection.StratifiedShuffleSplit(n_splits=5,test_size=0.1) # (45+45+45) vs. (5+5+5)
model=svm.SVC(C=10)
acc=[]
for train_index, test_index in sss.split(X, Y): # 5-fold cross-validation
    X_train, X_test = X[train_index], X[test_index]
    Y_train, Y_test = Y[train_index], Y[test_index]
    model.fit(X_train[:,0:2],Y_train) #training
    acc.append(model.predict(X_test[:,0:2])==Y_test) # testing
print(np.mean(acc))

0.8266666666666667


### 1.1 Tuning your ANN (3 points)
Tune your model hyperparameters (# of layers, # of units in each layer, activation function, optimizer, epochs, batch_size, etc.) to see if you can push your ANN performance up to ~9x% correct.

In [63]:
from keras.models import Sequential
from keras.layers import Dense,Dropout
from keras.utils import to_categorical
import keras

model = Sequential()
model.add(Dense(units=3, activation='relu'))
model.add(Dense(units=8, activation='relu'))
model.add(Dense(units=6, activation='relu'))
model.add(Dense(units=7, activation='relu'))
model.add(Dense(units=5, activation='relu'))
model.add(Dense(units=8, activation='relu'))
model.add(Dense(units=10, activation='relu'))
model.add(Dense(units=3, activation='softmax'))
adam=keras.optimizers.Adam(lr=0.018)
model.compile(loss='categorical_crossentropy',optimizer=adam,metrics=['accuracy'])

acc=[]
for train_index, test_index in sss.split(X, Y): # 5-fold cross-validation
    X_train, X_test = X[train_index], X[test_index]
    Y_train, Y_test = Y[train_index], Y[test_index]
    model.fit(X_train[:,0:2], to_categorical(Y_train), epochs=400, batch_size=100)
    acc.append(argmax(model.predict(X_test[:,0:2]),1)==Y_test) # testing
print(np.mean(acc))

Epoch 1/400
135/135 [==============================] - 0s 2ms/step - loss: 1.1169 - accuracy: 0.3259
Epoch 2/400
135/135 [==============================] - 0s 93us/step - loss: 1.0983 - accuracy: 0.3333
Epoch 3/400
135/135 [==============================] - 0s 78us/step - loss: 1.1029 - accuracy: 0.3333
Epoch 4/400
135/135 [==============================] - 0s 72us/step - loss: 1.0974 - accuracy: 0.3407
Epoch 5/400
135/135 [==============================] - 0s 73us/step - loss: 1.0920 - accuracy: 0.3333
Epoch 6/400
135/135 [==============================] - 0s 70us/step - loss: 1.0947 - accuracy: 0.3333
Epoch 7/400
135/135 [==============================] - 0s 85us/step - loss: 1.0946 - accuracy: 0.3333
Epoch 8/400
135/135 [==============================] - 0s 83us/step - loss: 1.0915 - accuracy: 0.3333
Epoch 9/400
135/135 [==============================] - 0s 57us/step - loss: 1.0886 - accuracy: 0.3333
Epoch 10/400
135/135 [==============================] - 0s 61us/step - loss: 1.0850

135/135 [==============================] - 0s 93us/step - loss: 0.4077 - accuracy: 0.7852
Epoch 159/400
135/135 [==============================] - 0s 207us/step - loss: 0.4000 - accuracy: 0.7704
Epoch 160/400
135/135 [==============================] - 0s 198us/step - loss: 0.4037 - accuracy: 0.7407
Epoch 161/400
135/135 [==============================] - 0s 142us/step - loss: 0.4077 - accuracy: 0.7852
Epoch 162/400
135/135 [==============================] - 0s 152us/step - loss: 0.3981 - accuracy: 0.7926
Epoch 163/400
135/135 [==============================] - 0s 116us/step - loss: 0.4093 - accuracy: 0.7407
Epoch 164/400
135/135 [==============================] - 0s 161us/step - loss: 0.3945 - accuracy: 0.7926
Epoch 165/400
135/135 [==============================] - 0s 115us/step - loss: 0.4045 - accuracy: 0.7778
Epoch 166/400
135/135 [==============================] - 0s 137us/step - loss: 0.3889 - accuracy: 0.8074
Epoch 167/400
135/135 [==============================] - 0s 74us/step 

Epoch 314/400
135/135 [==============================] - 0s 197us/step - loss: 0.3790 - accuracy: 0.7926
Epoch 315/400
135/135 [==============================] - 0s 120us/step - loss: 0.3783 - accuracy: 0.8074
Epoch 316/400
135/135 [==============================] - 0s 68us/step - loss: 0.3792 - accuracy: 0.8074
Epoch 317/400
135/135 [==============================] - 0s 106us/step - loss: 0.3804 - accuracy: 0.7778
Epoch 318/400
135/135 [==============================] - 0s 101us/step - loss: 0.3804 - accuracy: 0.7926
Epoch 319/400
135/135 [==============================] - 0s 101us/step - loss: 0.3780 - accuracy: 0.8074
Epoch 320/400
135/135 [==============================] - 0s 82us/step - loss: 0.3797 - accuracy: 0.8074
Epoch 321/400
135/135 [==============================] - 0s 93us/step - loss: 0.3825 - accuracy: 0.7926
Epoch 322/400
135/135 [==============================] - 0s 116us/step - loss: 0.3769 - accuracy: 0.8000
Epoch 323/400
135/135 [==============================] - 0

135/135 [==============================] - 0s 104us/step - loss: 0.3670 - accuracy: 0.8074
Epoch 71/400
135/135 [==============================] - 0s 56us/step - loss: 0.3560 - accuracy: 0.8148
Epoch 72/400
135/135 [==============================] - 0s 56us/step - loss: 0.3635 - accuracy: 0.8148
Epoch 73/400
135/135 [==============================] - 0s 105us/step - loss: 0.3653 - accuracy: 0.8074
Epoch 74/400
135/135 [==============================] - 0s 115us/step - loss: 0.3613 - accuracy: 0.8000
Epoch 75/400
135/135 [==============================] - ETA: 0s - loss: 0.3649 - accuracy: 0.82 - 0s 128us/step - loss: 0.3568 - accuracy: 0.8222
Epoch 76/400
135/135 [==============================] - 0s 201us/step - loss: 0.3557 - accuracy: 0.8074
Epoch 77/400
135/135 [==============================] - 0s 198us/step - loss: 0.3591 - accuracy: 0.8074
Epoch 78/400
135/135 [==============================] - 0s 142us/step - loss: 0.3664 - accuracy: 0.8222
Epoch 79/400
135/135 [===============

135/135 [==============================] - 0s 37us/step - loss: 0.3703 - accuracy: 0.8222
Epoch 226/400
135/135 [==============================] - 0s 36us/step - loss: 0.3648 - accuracy: 0.8296
Epoch 227/400
135/135 [==============================] - 0s 60us/step - loss: 0.3635 - accuracy: 0.8074
Epoch 228/400
135/135 [==============================] - 0s 127us/step - loss: 0.3634 - accuracy: 0.8074
Epoch 229/400
135/135 [==============================] - 0s 150us/step - loss: 0.3595 - accuracy: 0.8148
Epoch 230/400
135/135 [==============================] - 0s 73us/step - loss: 0.3746 - accuracy: 0.8148
Epoch 231/400
135/135 [==============================] - 0s 96us/step - loss: 0.3662 - accuracy: 0.8148
Epoch 232/400
135/135 [==============================] - 0s 129us/step - loss: 0.3604 - accuracy: 0.8296
Epoch 233/400
135/135 [==============================] - 0s 112us/step - loss: 0.3642 - accuracy: 0.8074
Epoch 234/400
135/135 [==============================] - 0s 135us/step - l

135/135 [==============================] - 0s 70us/step - loss: 0.3850 - accuracy: 0.8148
Epoch 382/400
135/135 [==============================] - 0s 87us/step - loss: 0.3760 - accuracy: 0.8148
Epoch 383/400
135/135 [==============================] - 0s 53us/step - loss: 0.3682 - accuracy: 0.8370
Epoch 384/400
135/135 [==============================] - 0s 70us/step - loss: 0.3717 - accuracy: 0.8222
Epoch 385/400
135/135 [==============================] - 0s 81us/step - loss: 0.3652 - accuracy: 0.8296
Epoch 386/400
135/135 [==============================] - 0s 56us/step - loss: 0.3674 - accuracy: 0.8148
Epoch 387/400
135/135 [==============================] - 0s 43us/step - loss: 0.3696 - accuracy: 0.8148
Epoch 388/400
135/135 [==============================] - 0s 68us/step - loss: 0.3636 - accuracy: 0.8296
Epoch 389/400
135/135 [==============================] - 0s 48us/step - loss: 0.3625 - accuracy: 0.8296
Epoch 390/400
135/135 [==============================] - 0s 51us/step - loss: 

135/135 [==============================] - 0s 80us/step - loss: 0.3849 - accuracy: 0.8074
Epoch 140/400
135/135 [==============================] - 0s 81us/step - loss: 0.3879 - accuracy: 0.7704
Epoch 141/400
135/135 [==============================] - 0s 65us/step - loss: 0.3856 - accuracy: 0.7852
Epoch 142/400
135/135 [==============================] - 0s 62us/step - loss: 0.3821 - accuracy: 0.8000
Epoch 143/400
135/135 [==============================] - 0s 76us/step - loss: 0.3883 - accuracy: 0.7704
Epoch 144/400
135/135 [==============================] - 0s 63us/step - loss: 0.3823 - accuracy: 0.7852
Epoch 145/400
135/135 [==============================] - 0s 62us/step - loss: 0.3887 - accuracy: 0.7852
Epoch 146/400
135/135 [==============================] - 0s 53us/step - loss: 0.4016 - accuracy: 0.7556
Epoch 147/400
135/135 [==============================] - 0s 59us/step - loss: 0.4049 - accuracy: 0.7778
Epoch 148/400
135/135 [==============================] - 0s 65us/step - loss: 

135/135 [==============================] - 0s 59us/step - loss: 0.3844 - accuracy: 0.7926
Epoch 296/400
135/135 [==============================] - 0s 46us/step - loss: 0.3836 - accuracy: 0.7926
Epoch 297/400
135/135 [==============================] - 0s 67us/step - loss: 0.3897 - accuracy: 0.7852
Epoch 298/400
135/135 [==============================] - 0s 55us/step - loss: 0.3845 - accuracy: 0.7926
Epoch 299/400
135/135 [==============================] - 0s 58us/step - loss: 0.3855 - accuracy: 0.7778
Epoch 300/400
135/135 [==============================] - 0s 62us/step - loss: 0.3883 - accuracy: 0.7630
Epoch 301/400
135/135 [==============================] - 0s 57us/step - loss: 0.3832 - accuracy: 0.8074
Epoch 302/400
135/135 [==============================] - 0s 50us/step - loss: 0.3961 - accuracy: 0.7852
Epoch 303/400
135/135 [==============================] - 0s 66us/step - loss: 0.3863 - accuracy: 0.8000
Epoch 304/400
135/135 [==============================] - 0s 55us/step - loss: 

Epoch 52/400
135/135 [==============================] - 0s 64us/step - loss: 0.3652 - accuracy: 0.8148
Epoch 53/400
135/135 [==============================] - 0s 58us/step - loss: 0.3623 - accuracy: 0.8222
Epoch 54/400
135/135 [==============================] - 0s 46us/step - loss: 0.3610 - accuracy: 0.8074
Epoch 55/400
135/135 [==============================] - 0s 58us/step - loss: 0.3616 - accuracy: 0.8074
Epoch 56/400
135/135 [==============================] - 0s 51us/step - loss: 0.3541 - accuracy: 0.8222
Epoch 57/400
135/135 [==============================] - 0s 58us/step - loss: 0.3559 - accuracy: 0.8222
Epoch 58/400
135/135 [==============================] - 0s 56us/step - loss: 0.3526 - accuracy: 0.8148
Epoch 59/400
135/135 [==============================] - 0s 61us/step - loss: 0.3553 - accuracy: 0.8296
Epoch 60/400
135/135 [==============================] - 0s 65us/step - loss: 0.3552 - accuracy: 0.8148
Epoch 61/400
135/135 [==============================] - 0s 53us/step - lo

135/135 [==============================] - 0s 59us/step - loss: 0.3527 - accuracy: 0.8296
Epoch 210/400
135/135 [==============================] - 0s 74us/step - loss: 0.3474 - accuracy: 0.8296
Epoch 211/400
135/135 [==============================] - 0s 78us/step - loss: 0.3469 - accuracy: 0.8370
Epoch 212/400
135/135 [==============================] - 0s 72us/step - loss: 0.3468 - accuracy: 0.8296
Epoch 213/400
135/135 [==============================] - 0s 64us/step - loss: 0.3469 - accuracy: 0.8370
Epoch 214/400
135/135 [==============================] - 0s 69us/step - loss: 0.3486 - accuracy: 0.8296
Epoch 215/400
135/135 [==============================] - 0s 58us/step - loss: 0.3571 - accuracy: 0.8148
Epoch 216/400
135/135 [==============================] - 0s 58us/step - loss: 0.3599 - accuracy: 0.8148
Epoch 217/400
135/135 [==============================] - 0s 95us/step - loss: 0.3504 - accuracy: 0.8296
Epoch 218/400
135/135 [==============================] - 0s 64us/step - loss: 

135/135 [==============================] - 0s 57us/step - loss: 0.3586 - accuracy: 0.8296
Epoch 366/400
135/135 [==============================] - 0s 51us/step - loss: 0.3600 - accuracy: 0.8148
Epoch 367/400
135/135 [==============================] - 0s 61us/step - loss: 0.3575 - accuracy: 0.8222
Epoch 368/400
135/135 [==============================] - 0s 57us/step - loss: 0.3601 - accuracy: 0.8222
Epoch 369/400
135/135 [==============================] - 0s 54us/step - loss: 0.3602 - accuracy: 0.8222
Epoch 370/400
135/135 [==============================] - 0s 60us/step - loss: 0.3587 - accuracy: 0.8222
Epoch 371/400
135/135 [==============================] - 0s 61us/step - loss: 0.3593 - accuracy: 0.8296
Epoch 372/400
135/135 [==============================] - 0s 58us/step - loss: 0.3591 - accuracy: 0.8148
Epoch 373/400
135/135 [==============================] - 0s 70us/step - loss: 0.3590 - accuracy: 0.8222
Epoch 374/400
135/135 [==============================] - 0s 71us/step - loss: 

Epoch 44/400
135/135 [==============================] - 0s 81us/step - loss: 0.3711 - accuracy: 0.8000
Epoch 45/400
135/135 [==============================] - 0s 87us/step - loss: 0.3779 - accuracy: 0.7852
Epoch 46/400
135/135 [==============================] - 0s 60us/step - loss: 0.3696 - accuracy: 0.8000
Epoch 47/400
135/135 [==============================] - 0s 55us/step - loss: 0.3715 - accuracy: 0.7778
Epoch 48/400
135/135 [==============================] - 0s 58us/step - loss: 0.3791 - accuracy: 0.7778
Epoch 49/400
135/135 [==============================] - 0s 49us/step - loss: 0.3666 - accuracy: 0.8148
Epoch 50/400
135/135 [==============================] - 0s 49us/step - loss: 0.3831 - accuracy: 0.7926
Epoch 51/400
135/135 [==============================] - 0s 70us/step - loss: 0.3821 - accuracy: 0.8000
Epoch 52/400
135/135 [==============================] - 0s 50us/step - loss: 0.3672 - accuracy: 0.8148
Epoch 53/400
135/135 [==============================] - 0s 51us/step - lo

135/135 [==============================] - 0s 70us/step - loss: 0.3613 - accuracy: 0.8148
Epoch 202/400
135/135 [==============================] - 0s 59us/step - loss: 0.3833 - accuracy: 0.7926
Epoch 203/400
135/135 [==============================] - 0s 52us/step - loss: 0.3552 - accuracy: 0.8148
Epoch 204/400
135/135 [==============================] - 0s 58us/step - loss: 0.3920 - accuracy: 0.7704
Epoch 205/400
135/135 [==============================] - 0s 53us/step - loss: 0.3900 - accuracy: 0.7704
Epoch 206/400
135/135 [==============================] - 0s 58us/step - loss: 0.3659 - accuracy: 0.8000
Epoch 207/400
135/135 [==============================] - 0s 64us/step - loss: 0.3689 - accuracy: 0.8148
Epoch 208/400
135/135 [==============================] - 0s 66us/step - loss: 0.3713 - accuracy: 0.8000
Epoch 209/400
135/135 [==============================] - 0s 62us/step - loss: 0.3620 - accuracy: 0.7926
Epoch 210/400
135/135 [==============================] - 0s 66us/step - loss: 

Epoch 357/400
135/135 [==============================] - 0s 76us/step - loss: 0.3648 - accuracy: 0.7926
Epoch 358/400
135/135 [==============================] - 0s 81us/step - loss: 0.3650 - accuracy: 0.8148
Epoch 359/400
135/135 [==============================] - 0s 88us/step - loss: 0.3721 - accuracy: 0.7852
Epoch 360/400
135/135 [==============================] - 0s 75us/step - loss: 0.3607 - accuracy: 0.8074
Epoch 361/400
135/135 [==============================] - 0s 56us/step - loss: 0.3734 - accuracy: 0.7778
Epoch 362/400
135/135 [==============================] - 0s 75us/step - loss: 0.3749 - accuracy: 0.7778
Epoch 363/400
135/135 [==============================] - 0s 75us/step - loss: 0.3628 - accuracy: 0.7926
Epoch 364/400
135/135 [==============================] - 0s 60us/step - loss: 0.3685 - accuracy: 0.7926
Epoch 365/400
135/135 [==============================] - 0s 67us/step - loss: 0.3740 - accuracy: 0.8074
Epoch 366/400
135/135 [==============================] - 0s 98us

### 1.2 Is your (deep) network better than SVM? Why or why not? (3 points)

#### network is better than SVM，deep network平均準確率0.8666666666666667>0.8266666666666667(SVM準確率)
#### =>雖然在較少極值的小數據集裡，SVM具有著優勢。然而 network可以讓我們建立充分逼近複雜的非線性關係，且具有聯想能力，容錯與嚕棒能力也較高
#### 但在試驗時，我仍然很難讓network準確率破90％（試過一、兩百多次，目前試過最高最確率是0.88但忘記存ㄌ），若SVM可以向題目所說準確率到8X%，那network不一定比SVM好，因為network可能會困在local best， 甚至有時候無法逼近到最佳解，且多少的layer等參數調整多少也是一個問題，此外資料集的合適性也有差，通常network需要資料集極大。